polaris-service/src/main/java/org/apache/polaris/service/persistence/MetadataCacheManager.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.polaris.core.PolarisCallContext;
+import org.apache.polaris.core.PolarisConfiguration;
+import org.apache.polaris.core.entity.PolarisEntity;
+import org.apache.polaris.core.entity.PolarisEntitySubType;
+import org.apache.polaris.core.entity.TableLikeEntity;
+import org.apache.polaris.core.persistence.PolarisMetaStoreManager;
+import org.apache.polaris.core.persistence.PolarisResolvedPathWrapper;
+import org.apache.polaris.core.persistence.resolver.PolarisResolutionManifestCatalogView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MetadataCacheManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetadataCacheManager.class);
+
+  /**
+   * Load the cached {@link Table} or fall back to `fallback` if one doesn't exist. If the metadata
+   * is not currently cached, it may be added to the cache.
+   */
+  public static TableMetadata loadTableMetadata(
+      TableIdentifier tableIdentifier,
+      long maxBytesToCache,
+      PolarisCallContext callContext,
+      PolarisMetaStoreManager metastoreManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView,
+      Supplier<TableMetadata> fallback) {
+    LOGGER.debug(String.format("Loading cached metadata for %s", tableIdentifier));
+    PolarisResolvedPathWrapper resolvedEntities =
+        resolvedEntityView.getResolvedPath(tableIdentifier, PolarisEntitySubType.TABLE);
+    TableLikeEntity tableLikeEntity = TableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+    boolean isCacheValid =
+        tableLikeEntity.getMetadataCacheLocation() != null
+            && tableLikeEntity
+                .getMetadataLocation()
+                .equals(tableLikeEntity.getMetadataCacheLocation());
+    if (isCacheValid) {
+      LOGGER.debug(String.format("Using cached metadata for %s", tableIdentifier));
+      return TableMetadataParser.fromJson(tableLikeEntity.getMetadataCacheContent());
+    } else {
+      TableMetadata metadata = fallback.get();
+      PolarisMetaStoreManager.EntityResult cacheResult =
+          cacheTableMetadata(
+              tableLikeEntity,
+              metadata,
+              maxBytesToCache,
+              callContext,
+              metastoreManager,
+              resolvedEntityView);
+      if (!cacheResult.isSuccess()) {
+        LOGGER.debug(String.format("Failed to cache metadata for %s", tableIdentifier));
+      }
+      return metadata;
+    }
+  }
+
+  /**
+   * Attempt to add table metadata to the cache
+   *
+   * @return The result of trying to cache the metadata
+   */
+  private static PolarisMetaStoreManager.EntityResult cacheTableMetadata(
+      TableLikeEntity tableLikeEntity,
+      TableMetadata metadata,
+      long maxBytesToCache,
+      PolarisCallContext callContext,
+      PolarisMetaStoreManager metaStoreManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView) {
+    String json = TableMetadataParser.toJson(metadata);
```

Comment:
TBH, this requires a lot of duplicate expensive operations:
1. it loads the metadata JSON from the object store
2. parses JSON into a TableMetadata
3. serializes the TableMetadata to the same string representation as in step 1
4. converts the `String` into a `byte[]` representation - just to check the length

The latter steps are likely performed for every access to a table-metadata that exceeds the METADATA_CACHE_MAX_BYTES_NO_CACHING setting, adding a _lot_ of new heap pressure.

Reply:
I'm concerned about this too, but it's only there for users who have a non-infinite and non-zero `METADATA_CACHE_MAX_BYTES`. Even still, the overhead is not small. Do you think just checking the string length is sufficient?